core/engine/src/object/property_map.rs

```diff
@@ -108,42 +126,170 @@ impl IndexedProperties {
 
     /// Inserts a property descriptor with the specified key.
     fn insert(&mut self, key: u32, property: PropertyDescriptor) -> bool {
-        let vec = match self {
-            Self::Sparse(map) => return map.insert(key, property).is_some(),
-            Self::Dense(vec) => {
-                let len = vec.len() as u32;
-                if key <= len
-                    && property.value().is_some()
-                    && property.writable().unwrap_or(false)
-                    && property.enumerable().unwrap_or(false)
-                    && property.configurable().unwrap_or(false)
-                {
-                    // Fast Path: continues array access.
-
-                    let mut value = property
-                        .value()
-                        .cloned()
-                        .expect("already checked that the property descriptor has a value field");
-
-                    // If the key is pointing one past the last element, we push it!
-                    //
-                    // Since the previous key is the current key - 1. Meaning that the elements are continuos.
-                    if key == len {
-                        vec.push(value);
-                        return false;
+        let mut map =
+            match self {
+                Self::DenseI32(vec) => {
+                    let len = vec.len() as u32;
+                    if key <= len
+                        && property.value().is_some()
+                        && property.writable().unwrap_or(false)
+                        && property.enumerable().unwrap_or(false)
+                        && property.configurable().unwrap_or(false)
+                    {
+                        // Fast Path: continues array access.
+                        let value = property.value().cloned().expect(
+                            "already checked that the property descriptor has a value field",
+                        );
+
+                        // If it can fit in a i32 and the truncated version is
+                        // equal to the original then it is an integer.
+                        let is_rational_integer =
+                            |n: f64| n.to_bits() == f64::from(n as i32).to_bits();
+
+                        let value = match value {
+                            JsValue::Integer(n) => n,
+                            JsValue::Rational(n) if is_rational_integer(n) => n as i32,
```

Comment:
Isn't this check also done when creating a `JsValue` from an f64? Why do we need to also do it here?

Reply:
[Sadly no](https://github.com/boa-dev/boa/blob/main/core/engine/src/value/conversions/mod.rs#L49-L56), I thought of implementing that, the problem is that nothing prevents the user from manually constructing an `JsValue::Rational` that can fit in a `i32` because `JsValue` is a public enum.

We need to hide implementation detail from the user, we can create a struct that encapsulates the enum, that way it can only be constructed through it's respective constructors. (had similar issue when trying to keep the invariant in #3450 for `JsStr`)

I encountered this type of issue many times with `JsValue` and had to add the checks, so implementing this could enable a lot of optimizations :)

Reply:
Could you open an issue for this? It could be a good first issue for someone.